Chore: improve errors

BridgeAR opened the pull request:
<!--
    ESLint adheres to the [JS Foundation Code of Conduct](https://js.foundation/community/code-of-conduct).
-->

**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[x] Other, please explain:

This improves the tests a bit by rewriting them to be independent of Node.js cores `assert` error messages, by switching to a better `isEqual` implementation (feature detecting the Node.js version) and by removing `assert.doesNotThrow`.

Fixes: #10111
<!--
    If the item you've checked above has a template, please paste the template questions below and answer them. (If this pull request is addressing an issue, you can just paste a link to the issue here instead.)
-->

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (https://eslint.org/docs/developer-guide/contributing/pull-requests)
    - Include tests for this change
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

platinumazure commented:
I'm confused. Does this actually address what was brought up in #10111?

And why are we removing `assert.doesNotThrow()`?

(cc @not-an-aardvark)

alberto commented on lib/testers/rule-tester.js:
```diff
@@ -396,10 +396,14 @@ class RuleTester {
+            if (util.isDeepStrictEqual) {
                 assert.deepStrictEqual(beforeAST, afterAST, "Rule should not modify AST.");
+
+            // Do not use `assert.deepStrictEqual` in Node.js 6 and 7 due to performance issues, see #6111.
```
Looks like `util.isDeepStrictEqual` was added in node 9, so it won't run in node 8 either, right?

BridgeAR replied:
Uh, yes. Seems like it was not backported. I thought that would be the case.

not-an-aardvark commented on lib/testers/rule-tester.js:
```diff
@@ -553,7 +557,7 @@ class RuleTester {
 
                         // Message was an unexpected type
-                        assert.fail(message, null, "Error should be a string, object, or RegExp.");
+                        assert.fail("Error should be a string, object, or RegExp.");
```
What versions of Node support `assert.fail` with one argument? Currently, we're supporting Node 4 on master (although we are dropping support for Node 4 and 5 on master in a few days).

BridgeAR replied:
Oh, I forgot about that again. You are right, I'll revert that change.

not-an-aardvark commented on lib/testers/rule-tester.js:
```diff
@@ -396,10 +396,14 @@ class RuleTester {
-                // Not using directly to avoid performance problem in node 6.1.0. See #6111
+            // Feature detect the Node.js implementation.
+            if (util.isDeepStrictEqual) {
                 assert.deepStrictEqual(beforeAST, afterAST, "Rule should not modify AST.");
```
I'm confused about this:

* It looks like we're using feature detection by checking for `util.isDeepStrictEqual`, but then we're never actually calling `util.isDeepStrictEqual`. Is that intentional?
* Why is feature detection necessary here? Is there an advantage to using Node's implementation when available, rather than always using `lodash.isEqual`? I'm not sure the behaviors of `lodash.isEqual` and `util.isDeepStrictEqual` are identical, so it seems like using feature detection could lead to cases where the behavior is slightly different depending on the Node version, which seems undesirable.

BridgeAR replied:
Well, in this case `util.isDeepStrictEqual` could also directly be used. About different comparisons in different versions: that is also the case when directly using the Node.js assert version as the implementation improved over time. I personally do not have a strong opinion about it. lodash was introduced because of a bad algorithm that got fixed a while ago in `deepStrictEqual`.

not-an-aardvark commented on tests/bin/eslint.js:
```diff
@@ -221,11 +221,8 @@ describe("bin/eslint.js", () => {
-                        "Cache file should contain valid JSON"
-                    );
+                    // "Cache file should contain valid JSON"
+                    JSON.parse(fs.readFileSync(CACHE_PATH, "utf8"));
```
I think it would be better to move the `assert.doesNotThrow` changes to a separate PR. In that PR, it would probably be a good idea to add it to the [linting configuration](https://github.com/eslint/eslint/blob/218ee57deb5eaec6c9df6fbf45d10673b51a2c2b/packages/eslint-config-eslint/default.yml#L93-L101) too, so that no one adds additional `assert.doesNotThrow` calls in the future.

not-an-aardvark commented on tests/lib/testers/rule-tester.js:
```diff
@@ -13,7 +13,7 @@
     EventEmitter = require("events"),
     RuleTester = require("../../../lib/testers/rule-tester"),
-    assert = require("chai").assert;
+    assert = require("assert");
```
Can we keep using `chai.assert` here? I think we we generally use `chai.assert` in tests. (We use the regular `assert` module in `rule-tester` to avoid needing to add a dependency.) `chai.assert` has a few advantages over `assert`, e.g. `chai.assert.throws(func, "error string")` tests against the error message as expected rather than creating a message for the assertion error.

BridgeAR replied:
Sure, I will change it back. I personally would love to change the `assert` implementation to also check for the error message but that would be probably to much of a breaking change.

not-an-aardvark commented on tests/lib/testers/rule-tester.js:
```diff
@@ -126,7 +123,16 @@ describe("RuleTester", () => {
                 ]
             });
-        }, /Bad var\.((.*==)|(.*strictEqual)).*Bad error message/);
+        }, err => {
```
Instead of using a function here, could we just provide the string as the second argument?

```js
assert.throws(() => {
    // ...
}, new nodeAssert.AssertionError({ /* ... */ }).message);
```

(This would work if we use `chai.assert.throws` rather than Node's `assert.throws`.)

BridgeAR replied:
That does not work. The reason is that the operator has to be detected from the thrown error, since that changed over time as well.

BridgeAR commented:
I just updated the PR and removed all changes to e.g. `assert.doesNotThrow`. One more thing about lodash and the core equality testing: they are mainly doing the same but Node core is probably a super tiny bit stricter and it has a faster algorithm for e.g. Map and Set equality in case the entries are objects.

platinumazure commented on lib/testers/rule-tester.js:
```diff
@@ -396,10 +396,10 @@ class RuleTester {
-                // Not using directly to avoid performance problem in node 6.1.0. See #6111
-                assert.deepStrictEqual(beforeAST, afterAST, "Rule should not modify AST.");
+            // Feature detect the Node.js implementation and use that if available.
+            if (util.isDeepStrictEqual && !util.isDeepStrictEqual(beforeAST, afterAST) || !lodash.isEqual(beforeAST, afterAST)) {
```
Could the first two conditions be parenthesized, just to avoid any ambiguity for future readers? (Yes, I know the order of operations is clear here, but just to be absolutely safe.)

```js
if ((util.isDeepStrictEqual && !util.isDeepStrictEqual(beforeAST, afterAST)) || !lodash.isEqual(beforeAST, afterAST)) {
```

Thanks!

BridgeAR replied:
Addressed.

alberto commented:
@BridgeAR can you fix the linting errors?

BridgeAR commented:
@alberto done

BridgeAR commented:
I just realized that I never ran the tests since I changed the code from the Node core `assert` module back to the `chai` one.

However, this implementation will not work with chai as they do not support functions as second argument (http://www.chaijs.com/api/assert/#method_throws). With chai, I am not sure how to get the error at all.

So I can only think of doing this "old school" with try catch... but that is not really nice.

not-an-aardvark commented:
Following up on https://github.com/eslint/eslint/pull/10182#discussion_r179211757, what if we feature-detected the operator too?

```js
// at the top of the file
const NODE_ASSERT_STRICT_EQUAL_OPERATOR = (() => {
  try {
    assert.strictEqual(1, 2);
  } catch (err) {
    return err.operator;
  }
  throw new Error("unexpected successful assertion");
})();

// in a test somewhere
assert.throws(() => {
    // ...
}, new nodeAssert.AssertionError({ /* ... */ operator: NODE_ASSERT_STRICT_EQUAL_OPERATOR }).message);
```

That would avoid the need to use a function as the matcher in `assert.throws`.

BridgeAR commented:
Another way is to generate two error messages (with two two different operators that Node.js core used so far) and to create a regular expression that accepts both of those messages... but that is also pretty ugly.

MylesBorins commented:
This is breaking the test suite on Node 10 is there a way to backport to ESlint 4.x?

not-an-aardvark commented:
Hi @MylesBorins, I don't think we have any plans right now to do another v4.x release unless a critical bug is found (we're mainly working on v5.x right now and we don't have the manpower to maintain multiple release lines). If we do do another release, I'll make sure this is included in it.
